refer/addbib.1
.\" Copyright (c) 1983 Regents of the University of California.
.\" All rights reserved.
.\"
.\" This code contains changes by
.\"      Gunnar Ritter, Freiburg i. Br., Germany, 2005. All rights reserved.
.\"
.\" Conditions 1, 2, and 4 and the no-warranty notice below apply
.\" to these changes.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowedgement:
.\" 	This product includes software developed by the University of
.\" 	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\" Copyright(C) Caldera International Inc. 2001-2002. All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\"   Redistributions of source code and documentation must retain the
.\"    above copyright notice, this list of conditions and the following
.\"    disclaimer.
.\"   Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"   All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"      This product includes software developed or owned by Caldera
.\"      International, Inc.
.\"   Neither the name of Caldera International, Inc. nor the names of
.\"    other contributors may be used to endorse or promote products
.\"    derived from this software without specific prior written permission.
.\"
.\" USE OF THE SOFTWARE PROVIDED FOR UNDER THIS LICENSE BY CALDERA
.\" INTERNATIONAL, INC. AND CONTRIBUTORS ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
.\" WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED. IN NO EVENT SHALL CALDERA INTERNATIONAL, INC. BE
.\" LIABLE FOR ANY DIRECT, INDIRECT INCIDENTAL, SPECIAL, EXEMPLARY, OR
.\" CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
.\" SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
.\" BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
.\" WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
.\" OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
.\" EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.\" Sccsid @(#)addbib.1b	1.4 (gritter) 12/12/05
.\"
.\"
.\"	from 4.3BSD-Tahoe addbib.1	6.1 (Berkeley) 4/29/85
.\"
.\".TH ADDBIB 1 "April 29, 1985"
.TH ADDBIB 1 "12/12/05" "Heirloom Documentation Tools" "BSD System Compatibility"
.UC 5
.SH NAME
addbib \- create or extend bibliographic database
.SH SYNOPSIS
\fBaddbib\fR
[\fB\-p\fI\ promptfile\fR] [\fB\-a\fR] \fIdatabase\fR
.SH DESCRIPTION
When this program starts up, answering ``y''
to the initial ``Instructions?'' prompt yields directions;
typing ``n'' or \s-2RETURN\s0 skips them.
.I Addbib
then prompts for various bibliographic fields,
reads responses from the terminal,
and sends output records to a
.I database.
A null response (just \s-2RETURN\s0) means to leave out that field.
A minus sign (\-) means to go back to the previous field.
A trailing backslash allows a field to be continued on the next line.
The repeating ``Continue?'' prompt allows the user
either to resume by typing ``y'' or \s-2RETURN\s0,
to quit the current session by typing ``n'' or ``q'',
or to edit the
.I database
with any system editor \fI(vi, ex, edit, ed).\fP
.PP
The
.B \-a
option suppresses prompting for an abstract;
asking for an abstract is the default.
Abstracts are ended with a \s-2CTRL\s0-d.
The
.B \-p
option causes
.I addbib
to use a new prompting skeleton, defined in
.I promptfile.
This file should contain prompt strings, a tab,
and the key-letters to be written to the
.I database.
.PP
The most common key-letters and their meanings are given below.
.I Addbib
insulates you from these key-letters,
since it gives you prompts in English,
but if you edit the bibliography file later on,
you will need to know this information.
.sp
.nf
	%A	Author's name
	%B	Book containing article referenced
	%C	City (place of publication)
	%D	Date of publication
	%E	Editor of book containing article referenced
	%F	Footnote number or label (supplied by \fIrefer\fP\|)
	%G	Government order number
	%H	Header commentary, printed before reference
	%I	Issuer (publisher)
	%J	Journal containing article
	%K	Keywords to use in locating reference
	%L	Label field used by \fB\-k\fP option of \fIrefer\fP
	%M	Bell Labs Memorandum (undefined)
	%N	Number within volume
	%O	Other commentary, printed at end of reference
	%P	Page number(s)
	%Q	Corporate or Foreign Author (unreversed)
	%R	Report, paper, or thesis (unpublished)
	%S	Series title
	%T	Title of article or book
	%V	Volume number
	%X	Abstract \(em used by \fIroffbib\fP, not by \fIrefer\fP
	%Y,Z	ignored by \fIrefer\fP
.fi
.sp
Except for `A', each field should be given just once.
Only relevant fields should be supplied.
An example is:
.sp
.nf
	%A	Bill Tuthill
	%T	Refer \(em A Bibliography System
	%I	Computing Services
	%C	Berkeley
	%D	1982
	%O	\s-1UNX\s0 4.3.5.
.fi
.sp
.SH FILES
.DT
promptfile	optional file to define prompting
.SH SEE ALSO
refer(1), sortbib(1), roffbib(1), indxbib(1), lookbib(1)
.SH NOTES
Written by Al Stangenberger & Bill Tuthill.


CHANGES
2014-10-29
    Documentation revised by Pierre-Jean Fichet:  doc/fonts/
    changed to use free available OTF fonts.

2014-10-27
    col(1) bugfix: Avoid to access chars before string.

    troff: fdprintf changed back to libc fprintf.

    troff: 32 bit bug in register and macro allocation fixed (Bug
    reported by Anthony J. Bentley).

    nroff/troff made -Wall clean.

2014-10-22
    OpenSolaris col(1) added.

2014-10-06
    Drawing of filled polygone (\D'P ...'), filled circle (\D'C
    ...') and filled ellipse (\D'E ...') added.

2014-09-21
    *roff: Option -V for version information added.

2014-09-14
    Any version of roff (e.g. UNIX V7) does count \} as a macro
    argument regarding \n(.$.  This makes \n(.$ rather random if
    you cannot force users to put \} on a separate line *or* to
    put it behind the macro call and substract 1 from \n(.$.  To
    make \n(.$ meaningful again this bug has been fixed now, i.e.
    \} is not considered as a macro argument anymore.

2014-09-07
    Documentation revised by Pierre-Jean Fichet:  doc/just/ and
    doc/quickstart/ changed to use free available OTF fonts.

2014-09-03
    Bug reported by Pierre-Jean Fichet fixed:  If character '"'
    is used in a .char request it did not work in a .if request
    as the string compare delimiter.

2014-08-28
    Bug reported by Blake McBride fixed:  Very old versions of
    the mm macros inserted cut marks in the document for
    typesetters which had used endless paper rolls.  Although
    they are only of historical significance they had never been
    removed in the Solaris version of mm.  They are now switched
    of (via .rm )k).

2014-08-23
    Bugfix by Steffen Nurpmeso:  In .rm code a function return
    value is unchecked used as a pointer.

2014-08-20
    eqn(1) did not compile because of bison to BSD yacc differ-
    ences.  Sorry, had not tested on Linux this time...

2014-08-18
    Experimental bugfix: When .bp is directly followed by .. and
    .bp triggers a macro which calls a macro then document text
    is printed at that trap position.  This bug is also in other
    ditroff derivatives (e.g. DWB and Plan9).  It is difficult to
    track the bug to it's root cause so it is an experimental fix
    at the moment.

2014-08-17
    \fP after \fS did set the font before the previous font and
    not the previous one.

2014-08-16
    Bug reported by Pierre-Jean Fichet fixed:  If character "!"
    is used in a .char request it did not work in a .if request
    as the NOT operator.

2014-08-14
    The eqn parser used type int for all token values--even
    pointers.  This may work fine on 32 bit hardware but causes
    problems on 64 bit.  A new data type had been introduced to
    separate int tokens from pointers.

    Also a bug in the parser regarding movements (e.g. FWD) had
    been fixed.

2014-07-27
    A variable had been initilized after use.  This variable
    minus a literal value is used as an array index.  When data
    from the array is read an invalid memory segment is accessed.
    This had caused crashes on at least OpenBSD.


refer/addbib.c
/*	Copyright (c) 1984, 1986, 1987, 1988, 1989 AT&T	*/
/*	  All Rights Reserved  	*/


/*
 * Copyright (c) 1980 Regents of the University of California.
 * All rights reserved. The Berkeley software License Agreement
 * specifies the terms and conditions for redistribution.
 */

/*
 * Copyright (c) 1983, 1984 1985, 1986, 1987, 1988, Sun Microsystems, Inc.
 * All Rights Reserved.
 */

/*	from OpenSolaris "addbib.c	1.5	05/06/02 SMI" 	*/

/*
 * Portions Copyright (c) 2005 Gunnar Ritter, Freiburg i. Br., Germany
 *
 * Sccsid @(#)addbib.c	1.3 (gritter) 10/22/05
 */


#include <locale.h>
#include <stdio.h>
#include <ctype.h>
#include <signal.h>
#include <sys/wait.h>
#include <unistd.h>
#include <string.h>
#include <stdlib.h>
#define MAXENT 50

static struct skeleton {
	char prompt[20];	/* prompt user for entry */
	char keylet[5]; 	/* key letter for database */
} bibskel[MAXENT] = {
	{ "   Author:",	"%A" },
	{ "    Title:",	"%T" },
	{ "  Journal:",	"%J" },
	{ "   Volume:",	"%V" },
	{ "    Pages:",	"%P" },
	{ "Publisher:",	"%I" },
	{ "     City:",	"%C" },
	{ "     Date:",	"%D" },
	{ "    Other:",	"%O" },
	{ " Keywords:",	"%K" }
};

static int entries = 10;	/* total number of entries in bibskel */
static int abstract = 1;	/* asking for abstracts is the default */

static void usage(void);
static void addbib(FILE *, char *);
static void trim(char *);
static void bibedit(FILE *, char *, char *);
static void instruct(void);
static void rd_skel(char *);

static void
usage(void)			/* print proper usage and exit */
{
	puts("Usage:  addbib [-p promptfile] [-a] database\n\
\t-p: the promptfile defines alternate fields\n\
\t-a: don't include prompting for the abstract\n");
	exit(1);
}

int
main(int argc, char **argv)	/* addbib: bibliography entry program */
{
	FILE *fp;
	int i;

	if (argc == 1)
	{
		puts("You must specify a bibliography file (database).");
		usage();
	}
	for (i = 1; argv[i][0] == '-'; i++)
	{
		if (argv[i][1] == 'p')
		{
			if (i >= argc - 2)
			{
				puts("Not enough arguments for -p option.");
				usage();
			}
			rd_skel(argv[++i]);
		}
		else if (argv[i][1] == 'a')
		{
			if (i >= argc - 1)
			{
				puts("No bibliofile specified after -a.");
				usage();
			}
			abstract = 0;
		}
		else  /* neither -p nor -a */
		{
			printf("Invalid command line flag: %s\n", argv[i]);
			usage();
		}
	}
	if (i < argc - 1)
	{
		puts("Too many arguments with no options.");
		usage();
	}
	if ((fp = fopen(argv[i], "a")) == NULL)
	{
		perror(argv[i]);
		exit(1);
	}
	addbib(fp, argv[i]);	/* loop for input */
	return 0;
}

static void
addbib(FILE *fp, char *argv)	/* add entries to a bibliographic database */
{
	char line[BUFSIZ];
	int i = 0, firstln;

	printf("Instructions? ");
	fgets(line, BUFSIZ, stdin);
	if (line[0] == 'y' || line[0] == 'Y')
		instruct();
	while (1)
	{
		putchar('\n');
		putc('\n', fp);
		for (i = 0; i < entries; i++)
		{
			printf("%s\t", bibskel[i].prompt);
			if (fgets(line, BUFSIZ, stdin) == NULL)
			{
				clearerr(stdin);
				break;
			}
			if (line[0] == '-' && line[1] == '\n')
			{
				i -= 2;
				if (i < -1)
				{
					printf("Too far back.\n");
					i++;
				}
				continue;
			}
			else if (line[strlen(line)-2] == '\\')
			{
				if (line[0] != '\\')
				{
					line[strlen(line)-2] = '\n';
					line[strlen(line)-1] = 0;
					trim(line);
					fprintf(fp, "%s %s",
						bibskel[i].keylet, line);
				}
				printf("> ");
				again:
				fgets(line, BUFSIZ, stdin);
				if (line[strlen(line)-2] == '\\')
				{
					line[strlen(line)-2] = '\n';
					line[strlen(line)-1] = 0;
					trim(line);
					fputs(line, fp);
					printf("> ");
					goto again;
				}
				trim(line);
				fputs(line, fp);
			}
			else if (line[0] != '\n')
			{
				trim(line);
				fprintf(fp, "%s %s", bibskel[i].keylet, line);
			}
		}
		if (abstract)
		{
			puts(" Abstract: (ctrl-d to end)");
			firstln = 1;
			while (fgets(line, BUFSIZ, stdin))
			{
				if (firstln && line[0] != '%')
				{
					fprintf(fp, "%%X ");
					firstln = 0;
				}
				fputs(line, fp);
			}
			clearerr(stdin);
		}
		fflush(fp);	/* write to file at end of each cycle */
		if (ferror(fp))
		{
			perror(argv);
			exit(1);
		}
		editloop:
		printf("\nContinue? ");
			fgets(line, BUFSIZ, stdin);
		if (line[0] == 'e' || line[0] == 'v')
		{
			bibedit(fp, line, argv);
			goto editloop;
		}
		if (line[0] == 'q' || line[0] == 'n')
			return;
	}
}

static void
trim(char *line)		/* trim line of trailing white space */
{
	int n;

	n = strlen(line);
	while (--n >= 0)
	{
		if (!isspace(line[n]))
			break;
	}
	line[++n] = '\n';
	line[++n] = 0;
}

static void
bibedit(FILE *fp, char *cmd, char *arg)	/* edit database with edit, ex, or vi */
{
	int i = 0, status, pid;

	fclose(fp);
	while (!isspace(cmd[i]))
		i++;
	cmd[i] = 0;
	if ((pid = fork()) == 0)
	{
		if (cmd[0] == 'v' && cmd[1] == 'i')
			execlp(cmd, cmd, "+$", arg, NULL);
		else /* either ed, ex, or edit */
			execlp(cmd, cmd, arg, NULL);
	}
	signal(SIGINT, SIG_IGN);
	signal(SIGQUIT, SIG_IGN);
	while (wait(&status) != pid);
	signal(SIGINT, SIG_DFL);
	signal(SIGQUIT, SIG_DFL);
	if ((fp = fopen(arg, "a")) == NULL)
	{
		perror(arg);
		exit(1);
	}
}

static void
instruct(void)		/* give user elementary directions */
{
	putchar('\n');
	puts("Addbib will prompt you for various bibliographic fields.\n"
"If you don't need a particular field, just hit RETURN,\n"
"\tand that field will not appear in the output file.\n"
"If you want to return to previous fields in the skeleton,\n"
"\ta single minus sign will go back a field at a time.\n"
"\t(This is the best way to input multiple authors.)\n"
"If you have to continue a field or add an unusual field,\n"
"\ta trailing backslash will allow a temporary escape.\n"
"Finally, (without -a) you will be prompted for an abstract\n"
"Type in as many lines as you need, and end with a ctrl-d.\n"
"To quit, type `q' or `n' when asked if you want to continue.\n"
"To edit the database, type `edit', `vi', or `ex' instead.");

}

static void
rd_skel(char *arg)		/* redo bibskel from user-supplied file */
{
	FILE *pfp;
	char str[BUFSIZ];
	int entry, i, j;

	if ((pfp = fopen(arg, "r")) == NULL)
	{
		fprintf(stderr, "Promptfile ");
		perror(arg);
		exit(1);
	}
	for (entry = 0; fgets(str, BUFSIZ, pfp); entry++)
	{
		for (i = 0; str[i] != '\t' && str[i] != '\n'; i++)
			bibskel[entry].prompt[i] = str[i];
		bibskel[entry].prompt[i] = 0;
		if (str[i] == '\n')
		{
			fprintf(stderr, "No tabs between promptfile fields.\n");
			fprintf(stderr, "Format: prompt-string <TAB> %%key\n");
			exit(1);
		}
		for (i++, j = 0; str[i] != '\n'; i++, j++) 
			bibskel[entry].keylet[j] = str[i];
		bibskel[entry].keylet[j] = 0;

		if (entry >= MAXENT)
		{
			fprintf(stderr, "Too many entries in promptfile.\n");
			exit(1);
		}
	}
	entries = entry;
	fclose(pfp);
}
